[CELEBORN-823][DOC] Add Celeborn architecture document (#1746)
apache/incubator-celeborn: docs/developers/overview.md

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
---

# Celeborn Architecture

This article introduces high level Apache Celeborn(Incubating) Architecture. For more detailed description of each module/process,
please refer to dedicated articles.

## Why Celeborn
In distributed compute engines, data exchange between compute nodes is common but expensive. The cost comes from
the disk and network inefficiency (M * N between Mappers and Reducers) in traditional shuffle frame, as following:

![ESS](/assets/docs/ess.svg)

Besides inefficiency, traditional shuffle framework requires large local storage in compute node to store shuffle
data, thus blocks the adoption of disaggregated architecture.

Apache Celeborn(Incubating) solves the problems by reorganizing shuffle data in a more efficient way, and storing the data in
a separate service. The high level architecture of Celeborn is as follows:

> ![RSS](/assets/docs/rss.svg)

Reviewer: nit: rss.svg -> celeborn.svg
```suggestion
![Celeborn](/assets/docs/celeborn.svg)
```

---
license: |
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
  http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
---

# Celeborn Architecture

> This articles introduces high level Celeborn Architecture. For more detailed description of each module/process,

Reviewer 2: ```suggestion
This article introduces high level Apache Celeborn(Incubating) Architecture. For more detailed description of each module/process,
```

with the same `partitionId` in a logical `PartitionLocation`.

In normal cases each `PartitionLocation` corresponds to a single file. When a reducer requires for the partition's data,
it just needs one network connection and sequentially read the coarse grained file.

In abnormal cases, such as when the file grows too large, or push data fails, Celeborn spawns a new split of the
`PartitionLocation`, and future data within the partition will be pushed to the new split.

Client keeps the split information and tells reducer to read from all splits of the `PartitionLocation` to guarantee
no data is lost.

## Data Storage
Celeborn stores shuffle data in configurable multiple layers, i.e. `Memroy`, `Local Disks`, `Distributed File System`,
and `Object Store`. Users can specify any combination of the layers on each Worker.

Currently, Celeborn only supports `Local Disks` and `HDFS`. Supporting for other storage systems are under working.

## Compute Engine Integration
Celeborn's primary components(i.e. Master, Worker, Client) are engine irrelevant. The Client APIs are extensible
and easy to implement plugins for various engines.

> Currently, Celeborn officially supports Spark(both Spark 2.x and Spark 3.x), Flink(1.14/1.15/1.17), and

Reviewer 3: Also add a link to spark and flink? such as [Spark](https://spark.org.apache)

[Gluten](https://github.com/oap-project/gluten). Also developers are integrating Celeborn with other engines,

> for example [MR3](https://github.com/mr3project/hive-mr3).

Reviewer 2: Maybe this link is more accurate

```suggestion
for example [MR3](https://mr3docs.datamonad.com/docs/mr3/).
```

Celeborn community is also working on integrating Celeborn with other engines.

> ## Graceful Upgrade

Reviewer 2: Maybe "rolling upgrade" or "graceful shutdown"?

> In order not to impact running applications when upgrading Celeborn Cluster, Celeborn implements Graceful Upgrade.

Reviewer 3: Add a link to  `Graceful Upgrade`.?
